compiler/noirc_evaluator/src/ssa/checks/check_for_underconstrained_values.rs

```diff
@@ -32,6 +35,32 @@ impl Ssa {
             })
             .collect()
     }
+
+    /// Detect Brillig calls left unconstrained with manual asserts
+    /// and return a vector of bug reports if any have been found
+    pub(crate) fn check_for_missing_brillig_constraints(&mut self) -> Vec<SsaReport> {
+        // Skip the check if there are no Brillig functions involved
+        if !self.functions.values().any(|func| func.runtime().is_brillig()) {
```

Comment:
Can this be skipped since we're filtering for Brillig functions later on? If the pass is fairly performant, I'd expect this extra run through all functions to be slower than attempting the pass and finding zero Brillig functions.

Reply:
The pass can be slow sometimes since it's going over every instruction sometimes gathering quite a footprint for brillig calls, especially in case they are unrolled large and brillig calls are in fact left unchecked. This filter doesn't get into function bodies at all, it just checks if any of involved functions is a brillig. 

By the way, this also should be ensuring the check doesn't do anything beyond this in the aforementioned `unconditional_recursion` test even if it the execution gets there somehow, as there are no brillig functions in the test.

Reply:
Although yes, in case there are no brillig calls happening at all it would be a simple no-op run through every instruction. If you think it's alright to do it (that certainly doesn't take much time anyway), we can get rid of this initial check.

Reply:
I don't think this makes much difference either way tbh. The only situation where I see this causing a performance regression would be if someone wrote a program with a) no brillig calls (so the pass is should be skipped)  b) The program makes heavy use of folding so that there's lots of ACIR functions still in SSA at this point.

I think this is a very unlikely case as if the're a program which uses an advanced feature like folding they're likely to find _some_ opportunity for brillig calls to optimise their program. It's also a very cheap check to run on each function so I don't see a good reason not to.

Reply:
That said, I think a complete lack of brillig usage in a program which experiences significant slowdown due to this pass would be odd.

Reply:
Leaving this in then. @michaeljklein agreed?